### PR TITLE
fix #4894 - Model results data using pydantic

### LINKED ISSUE #4894: Redesign results API
The results API is a tad messy code to read and modify. It needs a bit of a redesign to make it more readable and maintainable.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-1105)
┆Issue Number: EXP-1105

### PR DESCRIPTION
**Pull request by emtwo**
This is still in progress. Tests are passing and some subset of data is modeled so I'm putting it up as a WIP, but more work needs to be done

### CODE REVIEW

**Comment by jaredlockhart** [comment]
👏 👏 👏 👏 👏

File: app/experimenter/visualization/api/v3/views.py [hunk 1]
```diff
@@ -36,6 +38,60 @@ class Statistic:
...
+        return iter(self.__root__)
+
+    def append(self, item):
+        self.__root__.append(item)
+
+    def extend(self, item):
+        self.__root__.extend(item)
+
+    def dict(self, exclude_none):
+        return [item.dict(exclude_none=exclude_none) for item in self.__root__]
```
Comment (jaredlockhart) [inline app/experimenter/visualization/api/v3/views.py:65]: Huh interesting.... you can't just extend `List` as the base class for this? Or does it need to extend BaseModel?  Maybe you can mix them together? `(List, BaseModel)`?

File: app/experimenter/visualization/api/v3/views.py [hunk 2]
```diff
@@ -36,6 +38,60 @@ class Statistic:
...
+class BranchComparisonData(BaseModel):
+    all: List[DataPoint] = []
+    first: DataPoint = DataPoint()
+
+
+class MetricData(BaseModel):
+    absolute: BranchComparisonData
+    difference: BranchComparisonData
+    relative_uplift: BranchComparisonData
+    significance: dict
```
Comment (jaredlockhart) [inline app/experimenter/visualization/api/v3/views.py:91]: Can you define this in a little more detail?  Like do you know the keys up front?  If you know the keys you can use [TypedDict](https://www.python.org/dev/peps/pep-0589/)

File: app/experimenter/visualization/api/v3/views.py [hunk 3]
```diff
@@ -112,55 +166,29 @@ def get_other_metrics_names_and_map(data, RESULTS_METRICS_MAP):
...
-                "branch": branch_name,
-                "point": round(branch_user_count / total_population * 100),
-            }
+            JetstreamDataPoint(
+                metric=Metric.USER_COUNT,
+                statistic=Statistic.PERCENT,
+                branch=branch_name,
+                point=round(branch_user_count / total_population * 100),
+            )
         )
```
Comment (jaredlockhart) [inline app/experimenter/visualization/api/v3/views.py:182]: Is there a way to move all this parsing/transforming logic into more compact methods right on the pydantic models?

File: app/experimenter/visualization/tests/api/constants.py [hunk 4]
```diff
@@ -1,371 +1,204 @@
...
+        DIFFERENCE_METRIC_DATA_OVERALL_NEGATIVE = cls.get_difference_metric_data(
+            DATA_POINT_D, {}, {"1": Significance.NEGATIVE}, is_retention=True
+        )
+
+        EMPTY_METRIC_DATA = MetricData(
+            absolute=BranchComparisonData(),
+            difference=BranchComparisonData(),
+            relative_uplift=BranchComparisonData(),
+            significance={"overall": {}, "weekly": {}},
+        )
```
Comment (jaredlockhart) [inline app/experimenter/visualization/tests/api/constants.py:136]: This is already such an improvement 👏 👏 👏 👏 👏

File: app/experimenter/visualization/tests/api/constants.py [hunk 5]
```diff
@@ -1,371 +1,204 @@
...
+        VARIANT_DATA_DEFAULT_METRIC_ROW_BINOMIAL.metric = "another_count"
+        VARIANT_DATA_DEFAULT_METRIC_ROW_BINOMIAL.statistic = "binomial"
+        VARIANT_DATA_DEFAULT_METRIC_ROW_BINOMIAL.branch = "variant"
+
+        VARIANT_POSITIVE_SIGNIFICANCE_DATA_ROW = DATA_IDENTITY_ROW.copy()
+        VARIANT_POSITIVE_SIGNIFICANCE_DATA_ROW.comparison = "difference"
+        VARIANT_POSITIVE_SIGNIFICANCE_DATA_ROW.metric = "search_count"
+        VARIANT_POSITIVE_SIGNIFICANCE_DATA_ROW.statistic = "mean"
+        VARIANT_POSITIVE_SIGNIFICANCE_DATA_ROW.branch = "variant"
+
```
Comment (jaredlockhart) [inline app/experimenter/visualization/tests/api/constants.py:67]: Maybe some of these strings like `"mean"` `"binomial"` etc could be constants so it's clear which values are arbitrary data like branch names and which are expected constants?

**Review by jaredlockhart — COMMENTED** [review]
Great job @emtwo this is such a huge improvement 🎉 🎉 🎉 🎉 🎉 

Couple ideas for possible enhancements but if you wanna do them later as followups and/or not at all I'm happy to already land this let me know what you think 👍

File: app/experimenter/visualization/api/v3/views.py [hunk 6]
```diff
@@ -36,6 +38,60 @@ class Statistic:
...
 
+class JetstreamDataPoint(BaseModel):
+    lower: int = None
+    upper: int = None
+    point: int = None
+    metric: str = None
+    branch: str = None
+    statistic: str = None
+    window_index: str = None
+    comparison: str = None
```
Comment (jaredlockhart) [inline app/experimenter/visualization/api/v3/views.py:49]: Oh and also maybe move all these data types out into their own file so this file is just for the actual views?

File: app/experimenter/visualization/api/v3/views.py (hunk 1 above)
Comment (emtwo) [reply app/experimenter/visualization/api/v3/views.py:65]: It won't let me do that, sadly. `TypeError: multiple bases have instance lay-out conflict`. Sounds like the two base classes do not cooperate

File: app/experimenter/visualization/api/v3/views.py (hunk 2 above)
Comment (emtwo) [reply app/experimenter/visualization/api/v3/views.py:91]: ah yes. missed doing this one

**Review by jaredlockhart — APPROVED** [review]
This is such a huge improvement @emtwo fantastic work 🎉 🎉 🎉 🎉 🎉

**Pull request merged**